### PR TITLE
Push notifications in nodejs using service worker

### LINKED ISSUE #2047: Push Notifications in Nodejs using Service Workers
### Proposed title of article
Push Notifications in Nodejs using Service Workers

### Introduction paragraph (2-3 paragraphs):
A push notification is a message that pops on the screen when another activity is being performed on a mobile device or a web application. In modern web applications, push notifications have a wider place. The notifications help to engage the user and draw the attention fully to a new activity that occurs in the web application. 

A service worker is a script run by the browser in the background separate from the web page and acts as a bridge between the browser and the network. It usually intercepts requests made by the document in the browser and redirects requests to cache memory to allow offline access.

Service workers can be used to implement push notifications in a web application. In this tutorial, I will explain a step-by-step method of implementing push notifications in a node application using service workers.

### Key takeaways: 
1. Understanding push notifications in web application.
2. Clear introduction to service workers in Nodejs.
3. Implementing push notifications in a Nodejs web app.

### References: 
https://developers.google.com/web/fundamentals/primers/service-workers

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by mercymeave**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #2047

### CODE REVIEW

File: content/articles/push-notification-in-nodejs-using-service-worker/index.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+### Goal
+In this article, I will explain a step-by-step implementation of push notifications using service workers in a Nodejs application.
+
+### Table of contents
+- Introduction
+- Prerequisites
+- Creating the node application
+- Building the server-side
+- Coding the client-side
+- Conclusion & Further Reading
```
Comment (paulodhiambo) [inline content/articles/push-notification-in-nodejs-using-service-worker/index.md:19]: Consider changing your table of content to:-

### Table of contents
- [Introduction](#introduction)
- [Goal](#goal)
- [Table of contents](#table-of-contents)
- [Prerequisites](#prerequisites)
- [Creating the Nodejs Application](#creating-the-nodejs-application)
  - [Init application](#init-application)
  - [Install dependencies](#install-dependencies)
- [Building the server side](#building-the-server-side)
  - [Import the dependencies](#import-the-dependencies)
  - [Generate VAPID Keys](#generate-vapid-keys)
  - [Setting the VAPID Keys with web push](#setting-the-vapid-keys-with-web-push)
  - [Creating the subscribe route](#creating-the-subscribe-route)
- [Coding the client side](#coding-the-client-side)
  - [Setting the static path](#setting-the-static-path)
  - [Creating the required files](#creating-the-required-files)
  - [Setting up the service worker](#setting-up-the-service-worker)
- [Run the server](#run-the-server)
- [Conclusion  and further reading](#conclusion--and-further-reading)

File: content/articles/push-notification-in-nodejs-using-service-worker/index.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+### Prerequisites
+- You will need to have Nodejs installed in your computer
+- A basic understanding of Nodejs
+- Code editor. I will use visual studio code
```
Comment (paulodhiambo) [inline content/articles/push-notification-in-nodejs-using-service-worker/index.md:24]: Provide download links to Node.js and Visual Studio code.

**Review by paulodhiambo — CHANGES_REQUESTED** [review]
@mercymeave amazing article. Review the requested changes then ping me when done.

**Comment by paulodhiambo** [comment]
@mercymeave amazing topic on Node.js push notifications. @hectorkambow you can give it a final review.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by paulodhiambo** [comment]
Hello :wave:
As you know, we were working on a new structure for our EngEd Repo, which has led to a slight slow down of article publishing. Everything is almost ready now.
In order to publish your article as soon as possible, please do the following:
- Download the most recent files off this PR (Make sure to pick the latest one with my revisions and edits).
- Delete your article directory from the *articles* folders.
- Do a git pull (take a look at the new structure of the repo after a git pull).
- Upload your article files to the [new *articles* folder inside *content*](https://github.com/section-engineering-education/engineering-education/tree/master/content/articles).

Ping me via Slack if you have any problems.

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

**Review by mercymeave — COMMENTED** [review]
Changes made

**Comment by hectorkambow** [comment]
Great work - very nice article - super easy to follow along 🚀 

Please be sure to go over the small minor edits I made for format mostly - some minor for flow 

We are getting ready to increase our volume of articles published monthly - and the best way to go about this is - reviewing articles like this one - that were well written and easy to read. 

👍  

cc @paulodhiambo

**Comment by mercymeave** [comment]
Thank you @hectorkambow and @paulodhiambo  for the review. I'll be sure to check the edits made.

**Pull request merged**